File: source/downloads/code/2012-02-14-simple-address-standardization/normalize-google.rb
require 'geocoder'
require 'bigdecimal'
require 'set'

module AddressNormalizer

  class NormalizedAddress
    Geocoder::Configuration.lookup = :google

    Attrs = [:address_line1, :address_line2, :city, :state_province,
             :postal_code, :country, :longitude, :formatted_address]

    attr_reader *Attrs

    def initialize(data)
      @address_line1  = nil
      @address_line2  = nil
      @city           = nil
      @postal_code    = nil
      @state_province = nil
      @country        = nil
      @latitude       = nil
      @longitude      = nil

      # The Google result consists of:
      #
      # - An array ("address_components") of hashes consisting of
      #   {"long_name" => "...", "short_name" => "...", "types" => [...]}
      # - A "geometry" hash, with the latitude and longitude
      # - A "partial_match" indicator (which we're ignoring)
      # - A "types" array (which we're also ignoring)

      data["address_components"].each do |hash|
        types = hash["types"]
        value = hash["long_name"]
        if types.include? "subpremise"
          @address_line2 = "##{value}"
        elsif types.include? "street_number"
          @house_number = value
        elsif types.include? "sublocality"
          @city = value
        elsif types.include? "locality"
          @city = value if @city.nil?
        elsif types.include? "country"
          @country = value
        elsif types.include? "postal_code"
          @postal_code = value
        elsif types.include? "route"
          @street = value
        elsif types.include? "administrative_area_level_1"
          @state_province = value
        end
      end

      @line1 = "#{@house_number} #{@street}"

      if data["formatted_address"]
        @formatted_address = data["formatted_address"]
      else
        @formatted_address = [
          @address_line1, @address_line2, @city, @state_province, @postal_code
        ].select {|s| ! (s.nil? || s.empty?)}.join(' ')
      end

      # Latitude and longitude

      if data["geometry"] and data["geometry"]["location"]
        loc = data["geometry"]["location"]
        @latitude = BigDecimal.new(loc["lat"].to_s)
        @longitude = BigDecimal.new(loc["lng"].to_s)
      end
    end

    def to_s
      @formatted_address
    end

    def inspect
      Hash[ Attrs.map {|field| [field, self.send(field)]} ]
    end
  end
  
  ACCEPTABLE_TYPES = Set.new(["street_address", "subpremise"])

  def normalize_street_address(raw_address_string)

    normalized_address = nil

    # Geocoder.search() returns an array of results. Take the first one.
    geocoded = Geocoder.search(raw_address_string)
    if geocoded && (geocoded.length > 0)
      # Geocoder returns data that may or may not be granular enough. For
      # instance, attempting to retrieve information about nonexistent
      # address '100 Main St, XYZ, PA, US' still returns a value, but the
      # value's type is "administrative_area_level_1", which means the data
      # is granular to a (U.S.) state. If it's a valid address, we should
      # get data that's more granular than that. Of the codes listed at
      # http://code.google.com/apis/maps/documentation/geocoding/#Types
      # we're interested in "street_address", "premise" and
      # "subpremise".
      data = geocoded[0].data
      types = Set.new(data["types"])
      if !(types & ACCEPTABLE_TYPES).empty?
        normalized_address = NormalizedAddress.new(data)
      end
    end

    normalized_address
  end
end

File: source/downloads/code/2012-02-14-simple-address-standardization/normalize_google.py
from googlemaps import GoogleMaps, GoogleMapsError
from decimal import Decimal

__all__ = ['NormalizedAddress', 'normalize_street_address']

class NormalizedAddress(object):

    def __init__(self, data):
        self.address_line1  = None
        self.address_line2  = None
        self.city           = None
        self.postal_code    = None
        self.state_province = None
        self.country        = None
        self.latitude       = None
        self.longitude      = None

        placemark = data['Placemark'][0]
        self.longitude, self.latitude = placemark['Point']['coordinates'][0:2]
        country_data = placemark['AddressDetails']['Country']
        details = country_data['AdministrativeArea']

        if details.get('SubAdministrativeArea') is not None:
            locality = details['SubAdministrativeArea']['Locality']
            self.city = locality['LocalityName']
        elif details.get('Locality') is not None:
            locality = details['Locality']
            self.city = locality['LocalityName']
        elif details.get('DependentLocality') is not None:
            locality = details['DependentLocality']
            self.city = locality['DependentLocalityName']
        else:
            raise GoogleMapsError(GoogleMapsError.G_GEO_UNKNOWN_ADDRESS)

        self.address_line1 = locality['Thoroughfare']['ThoroughfareName']

        # Break out any suite number, since it's not broken out in the data.
        try:
            i = self.address_line1.index('#')
            self.address_line2 = self.address_line1[i:].strip()
            self.address_line1 = self.address_line1[:i-1].strip()
        except ValueError:
            pass

        self.state_province = details['AdministrativeAreaName']
        self.postal_code = locality['PostalCode']['PostalCodeNumber']
        self.country = country_data['CountryName']
        self.formatted_address = placemark.get('AddressDetails', {}).get('address')
        if self.formatted_address is None:
            fields = [
                self.address_line1, self.address_line2, self.city,
                self.state_province, self.postal_code, self.country
            ]
            self.formatted_address = ' '.join([
                x for x in fields if (x is not None) and (x.strip() != '')
            ])

    def __str__(self):
        return self.formatted_address

    def __repr__(self):
        return str(self.__dict__)

def normalize_street_address(raw_address_string):

    gm = GoogleMaps() # Add API key for premium Google Maps service
    result = None

    try:
        data = gm.geocode(raw_address_string)
        if data is not None:
            result = NormalizedAddress(data)

    except KeyError:
        result = None

    except GoogleMapsError as ex:
        if ex.message != GoogleMapsError.G_GEO_UNKNOWN_ADDRESS:
            raise ex

    return result
